src/skagent/models/benchmarks.py

```diff
@@ -0,0 +1,1048 @@
+#!/usr/bin/env python3
+"""
+Complete Catalogue of Analytically Solvable Consumption-Savings Models
+
+This module implements the exhaustive collection of discrete-time consumption-savings
+dynamic programming problems for which the literature has succeeded in writing down
+true closed-form policies. The catalogue is complete as of June 13, 2025, to the best
+of our knowledge: no other analytically solvable DP variants are currently known.
+
+THEORETICAL FOUNDATION
+======================
+
+An entry qualifies for inclusion ONLY if:
+(i) The problem is a bona-fide dynamic programming problem
+(ii) The optimal c_t (and any other control) can be written in closed form
+    with no recursive objects left implicit
+
+Global Notation (Discrete Time Only)
+------------------------------------
+t ∈ {0,1,2,...}     : period index
+A_t                 : end-of-period assets (risk-free bond, gross return R=1+r>1)
+y_t                 : non-capital income
+c_t                 : consumption
+m_t := A_t + h_t    : cash-on-hand; h_t = E_t[∑_{s≥0} R^{-(s+1)} y_{t+s}]
+u(c)                : period utility
+TVC                 : lim_{T→∞} E_0[β^T u'(c_T) A_T] = 0
+"""
+
+from HARK.distributions import Normal, MeanOneLogNormal
+from skagent.model import Control, DBlock
+import torch
+import numpy as np
+from typing import Dict, Any, Callable
+
+# NUMERICAL TOLERANCE CONSTANTS
+# =============================
+EPS_STATIC = 1e-10  # Static identity verification (deterministic)
+EPS_EULER = 1e-8  # Euler equation residuals (stochastic)
+EPS_BUDGET = 1e-12  # Budget evolution (should be exact)
+EPS_VALIDATION = 1e-8  # General validation tolerance
+
+# DETERMINISTIC (PERFECT-FORESIGHT) BENCHMARKS
+# ============================================
+
+# D-1: Two-Period Log Utility
+# ---------------------------
+# The simplest possible consumption-savings problem: allocate wealth W between
+# consumption today (c₁) and consumption tomorrow (c₂), with log utility and
+# known interest rate R.
+
+# Mathematical formulation:
+#   max_{c₁,c₂} ln c₁ + β ln c₂
+#   s.t. c₁ + c₂/R = W
+
+# First-order conditions:
+#   1/c₁ = λ
+#   β/c₂ = λ/R
+
+# Combining: c₂ = βRc₁
+# Budget constraint: c₁ + βRc₁/R = W ⟹ c₁(1+β) = W
+
+# Closed-form solution:
+#   c₁ = W/(1+β)
+#   c₂ = βRW/(1+β)
+
+# This is the foundation for all intertemporal choice theory.
```

Comment:
not sure a comment like this is appropriate -- it's a bit disciplinary specific to economics. there might be intertermporal choice in non-economics fields that are about something else?